casper/src/main/scala/coop/rchain/casper/util/comm/FairRoundRobinDispatcher.scala

```diff
@@ -0,0 +1,144 @@
+package coop.rchain.casper.util.comm
+
+import scala.collection.immutable.Queue
+
+import cats.Show
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+
+import FairRoundRobinDispatcher._
+import coop.rchain.shared.Log
+
+class FairRoundRobinDispatcher[F[_]: Sync: Log, S: Show, M](
+    filter: M => F[Dispatch],
+    handle: (S, M) => F[Unit],
+    queue: Ref[F, Queue[S]],
+    messages: Ref[F, Map[S, Queue[M]]],
+    retries: Ref[F, Map[S, Int]],
+    skipped: Ref[F, Int],
+    maxSourceQueueSize: Int,
+    giveUpAfterSkipped: Int,
+    dropSourceAfterRetries: Int
+) {
+  assert(maxSourceQueueSize > 0)
+  assert(giveUpAfterSkipped >= 0)
+  assert(dropSourceAfterRetries >= 0)
+
+  def dispatch(source: S, message: M): F[Unit] =
+    filter(message).flatMap {
+      case Handle =>
+        ensureSourceExists(source) >>
+          enqueueMessage(source, message) >>
+          handleMessages(source)
+      case Pass => handle(source, message)
+      case Drop => ().pure
+    }
+
+  private[comm] def ensureSourceExists(source: S): F[Unit] =
+    messages.get
+      .map(_.contains(source))
+      .ifM(
+        ().pure,
+        queue.update(_.enqueue(source)) *>
+          messages.update(_ + (source -> Queue.empty[M])) *>
+          retries.update(_ + (source  -> 0)) *>
+          Log[F].debug(s"Added ${source.show} to the dispatch queue")
+      )
+
+  private[comm] def enqueueMessage(source: S, message: M): F[Unit] =
+    messages.get
+      .map(_(source).size < maxSourceQueueSize)
+      .ifM(
+        messages.update(ps => ps.updated(source, ps(source).enqueue(message))) *>
+          retries.update(_.updated(source, 0)),
+        Log[F].debug(s"Dropped message from ${source.show}")
+      )
+
+  private[comm] def handleMessages(source: S): F[Unit] =
+    queue.get.flatMap { q =>
+      if (q.head == source) ().pure
+      else failure(q.head)
+    } >> handleNextMessage // because maybe gave up on failure
+
+  private[comm] val handleNextMessage: F[Unit] =
+    queue.get.map(_.head) >>= { s =>
+      handleMessage(s).ifM(
+        success(s) >> handleNextMessage,
```

Comment:
It doesn't feel right - if `A` is at the head of the queue and enqueued 1k messages and a message from `B` arrives then it is going to wait till these 1k messages get handled. Or maybe there is something else at play?

Reply:
No, each `success` rotates the queue. So after the first message dispatched for `A` it's `B`'s turn.

Reply:
Ah, ok, now I see that this part is the key. Thanks for explaining.